Version API

BenMorganIO opened the pull request:
I'm thinking that `/api` should use the default version. That way when v2 is being made, we can let users use v1 and when Spree 4 is out, it will allow us to remove v1 and not have to namespace the controllers in v2.

BenMorganIO committed:
wrap api in a concern

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://benmorganio-version-api.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://benmorganio-version-api.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

BenMorganIO committed:
add v1 namespace

BenMorganIO committed:
there's no concern option in namespaces

BenMorganIO committed:
remove extra option values resource

BenMorganIO committed:
do end for the tCURLEYs

BenMorganIO commented:
47b3949 you passed on local!

JDutil commented on api/config/routes.rb:
```diff
@@ -1,56 +1,40 @@
-        put :generate_api_key
-        put :clear_api_key
-      end
+      put :generate_api_key, :clear_api_key, on: :member
```
personally I prefer the block

JDutil commented:
We should have some specs around the namespacing / defaults working as expected.

gmacdougall commented:
I'm in favour of this method of providing a versioned namespace. I would prefer that a pull request to change the formatting was separate from the one which changes functionality, as it makes it easier to understand what changed when.

I also concur with @JDutil that some specs around the routing would be nice to have.

BenMorganIO commented:
I'll split them up.

BenMorganIO commented:
Please see #6051.

BenMorganIO committed:
version spec /api

BenMorganIO committed:
namespace with a v1 folder

BenMorganIO committed:
wrap api controllers in v1 module

BenMorganIO committed:
don't version control the base_controller

BenMorganIO commented:
I though about changing the API so that it does some nice concerns and redirects, but when I was thinking about v2 and how much people want a new and improved API, I basically just decided to stop pulling tricks and legit namespace the controllers.

@JDutil and @gmacdougall how bad do think this move is?

BenMorganIO committed:
fix merge conflicts

BenMorganIO committed:
make the predicate change again

Didn’t make it in the merge fix

BenMorganIO commented:
I actually can't believe the specs are passing....

BenMorganIO committed:
add redirect to v1

@radar thanks!
http://stackoverflow.com/questions/9627546/api-versioning-for-rails-rout
es

BenMorganIO committed:
specs for the redirect

BenMorganIO committed:
move templates into v1 folder

BenMorganIO commented:
This change in the routes for versioning is the standard rails method. If anyone has any better ideas, please please please please please put them forward. Thanks :)

BenMorganIO committed:
address controller hounded

BenMorganIO committed:
images controller hounded

BenMorganIO commented on api/app/controllers/spree/api/v1/addresses_controller.rb:
```diff
@@ -0,0 +1,45 @@
+          end
+        end
+
+        private
```
done

BenMorganIO committed:
move api controller specs into v1 directory

BenMorganIO committed:
base controller spec belongs in api root

BenMorganIO committed:
add v1 namespace module to specs

BenMorganIO commented:
This attempt was a poor way to version the API.

Does anyone have any awesome plans of attack?

BenMorganIO committed:
update rabl templates to refer to v1 folder

BenMorganIO committed:
rabl cache spec uses v1

BenMorganIO committed:
have the controllers using the v1 templates

BenMorganIO committed:
use described class

BenMorganIO committed:
use v1 templates in specs

BenMorganIO committed:
update rabl cache spec to use v1 for products

BenMorganIO commented:
All spree_api specs passing now...

BenMorganIO committed:
checkouts controller hounded

BenMorganIO committed:
classifications controller hounded

BenMorganIO committed:
product properties controller hounded

BenMorganIO committed:
version routes in backend

This code worked before, but now there’s no redirect.

BenMorganIO committed:
use v1 route in taxonomies edit

JDutil commented:
I think namespacing and routing /api to v1 is fine.  It shouldn't be changing anyones behavior.  Let's see what people think, but I think this will be the best way forward.

JDutil commented:
8 Frontend specs are failing:
```
Building: frontend
Generating dummy Rails application...
Setting up dummy database...
/home/ubuntu/spree/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.0/lib/active_record/associations.rb:1693: warning: already initialized constant Spree::ShippingMethod::HABTM_Zones
/home/ubuntu/spree/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.0/lib/active_record/associations.rb:1693: warning: previous definition of HABTM_Zones was here
/home/ubuntu/spree/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.0/lib/active_record/associations.rb:1693: warning: already initialized constant Spree::ShippingMethod::HABTM_Zones
/home/ubuntu/spree/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.0/lib/active_record/associations.rb:1693: warning: previous definition of HABTM_Zones was here
Skipping installation no generator to run...
Instance method "open" is already defined in Object, use generic helper instead or set StateMachines::Machine.ignore_method_conflicts = true.
.............................................................................FF*..................................FF.......FFFF..........................................................

Pending:
  Address country requires state user changes to country without states required clears the state name
    # This is failing on the CI server, but not when you run the tests manually... It also does not fail locally on a machine.
    # ./spec/features/address_spec.rb:65

Failures:

  1) Address country requires state but has no state shows the state input field
     Failure/Error: expect(page).to have_selector(@state_name_css, visible: true)
       expected to find css "#order_bill_address_attributes_state_name" but there were no matches. Also found "", which matched the selector but not all filters.
     # ./spec/features/address_spec.rb:39:in `block (4 levels) in <top (required)>'

  2) Address country requires state and has state shows the state collection selection
     Failure/Error: expect(find(@state_select_css)['class']).to match(/required/)
       expected "form-control" to match /required/
       Diff:
       @@ -1,2 +1,2 @@
       -/required/
       +"form-control"
       
     # ./spec/features/address_spec.rb:56:in `block (4 levels) in <top (required)>'

  3) Checkout if coupon promotion, submits coupon along with payment makes sure payment reflects order total with discounts
     Failure/Error: click_on "Save and Continue"
     Capybara::Poltergeist::JavascriptError:
       One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).
       
       SyntaxError: Unable to parse JSON string
       SyntaxError: Unable to parse JSON string
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11591
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3144
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3256
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9312 in done
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9714
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9720
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9211
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11595
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4666
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4334
     # ./spec/features/checkout_spec.rb:381:in `block (3 levels) in <top (required)>'

  4) Checkout if coupon promotion, submits coupon along with payment invalid coupon doesnt create a payment record
     Failure/Error: click_on "Save and Continue"
     Capybara::Poltergeist::JavascriptError:
       One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).
       
       SyntaxError: Unable to parse JSON string
       SyntaxError: Unable to parse JSON string
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11591
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3144
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3256
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9312 in done
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9714
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9720
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9211
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11595
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4666
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4334
     # ./spec/features/checkout_spec.rb:390:in `block (4 levels) in <top (required)>'

  5) Coupon code promotions visitor makes checkout as guest without registration on the payment page informs about an invalid coupon code
     Failure/Error: click_button "Save and Continue"
     Capybara::Poltergeist::JavascriptError:
       One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).
       
       SyntaxError: Unable to parse JSON string
       SyntaxError: Unable to parse JSON string
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11591
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3144
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3256
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9312 in done
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9714
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9720
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9211
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11595
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4666
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4334
     # ./spec/features/coupon_code_spec.rb:56:in `block (4 levels) in <top (required)>'

  6) Coupon code promotions visitor makes checkout as guest without registration on the payment page informs the user about a coupon code which has exceeded its usage
     Failure/Error: click_button "Save and Continue"
     Capybara::Poltergeist::JavascriptError:
       One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).
       
       SyntaxError: Unable to parse JSON string
       SyntaxError: Unable to parse JSON string
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11591
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3144
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3256
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9312 in done
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9714
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9720
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9211
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11595
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4666
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4334
     # ./spec/features/coupon_code_spec.rb:65:in `block (4 levels) in <top (required)>'

  7) Coupon code promotions visitor makes checkout as guest without registration on the payment page can enter an invalid coupon code, then a real one
     Failure/Error: click_button "Save and Continue"
     Capybara::Poltergeist::JavascriptError:
       One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).
       
       SyntaxError: Unable to parse JSON string
       SyntaxError: Unable to parse JSON string
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11591
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3144
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3256
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9312 in done
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9714
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9720
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9211
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11595
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4666
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4334
     # ./spec/features/coupon_code_spec.rb:71:in `block (4 levels) in <top (required)>'

  8) Coupon code promotions visitor makes checkout as guest without registration on the payment page with a promotion applies a promotion to an order
     Failure/Error: click_button "Save and Continue"
     Capybara::Poltergeist::JavascriptError:
       One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).
       
       SyntaxError: Unable to parse JSON string
       SyntaxError: Unable to parse JSON string
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11591
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3144
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:3256
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9312 in done
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9714
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9720
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:9211
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:11595
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4666
           at http://127.0.0.1:42485/assets/spree/frontend/all.js:4334
     # ./spec/features/coupon_code_spec.rb:81:in `block (5 levels) in <top (required)>'

Finished in 3 minutes 11 seconds (files took 3.99 seconds to load)
185 examples, 8 failures, 1 pending

Failed examples:

rspec ./spec/features/address_spec.rb:34 # Address country requires state but has no state shows the state input field
rspec ./spec/features/address_spec.rb:50 # Address country requires state and has state shows the state collection selection
rspec ./spec/features/checkout_spec.rb:379 # Checkout if coupon promotion, submits coupon along with payment makes sure payment reflects order total with discounts
rspec ./spec/features/checkout_spec.rb:388 # Checkout if coupon promotion, submits coupon along with payment invalid coupon doesnt create a payment record
rspec ./spec/features/coupon_code_spec.rb:54 # Coupon code promotions visitor makes checkout as guest without registration on the payment page informs about an invalid coupon code
rspec ./spec/features/coupon_code_spec.rb:60 # Coupon code promotions visitor makes checkout as guest without registration on the payment page informs the user about a coupon code which has exceeded its usage
rspec ./spec/features/coupon_code_spec.rb:69 # Coupon code promotions visitor makes checkout as guest without registration on the payment page can enter an invalid coupon code, then a real one
rspec ./spec/features/coupon_code_spec.rb:79 # Coupon code promotions visitor makes checkout as guest without registration on the payment page with a promotion applies a promotion to an order
Finished running 2 projects(s) on node 1 / 3
- api SUCCESS

- frontend FAILURE ./build-ci.rb test returned exit code 1
```

BenMorganIO commented:
OK, jsut the frontend left :)

BenMorganIO committed:
refactor accurate_title

BenMorganIO committed:
version lookup context for order state

BenMorganIO committed:
user version 1 in spree.js

BenMorganIO commented:
OK, I think all the specs are passing now...

JDutil commented:
:beers:

BenMorganIO commented:
@JDutil something went wrong somewhere in the merge. I'm going to submit a second PR trying to fix the mess.
